### PR TITLE
DM-34253: Set up Source Injection package

### PR DESCRIPTION
**Pull request by leeskelvin**

### CODE REVIEW

**Comment by jmeyers314** [comment]
Looks like `ups/source_injection.table` needs at least

```
setupRequired(daf_butler)
setupRequired(pipe_base)
```

to run the tests.

File: .pre-commit-config.yaml
Comment (timj) [inline .pre-commit-config.yaml:18]: I think you need to run `pre-commit autoupdate` since 5.10.1 won't install any more.

Comment (jmeyers314) [reply .pre-commit-config.yaml:18]: How do you spell that?  I don't see `pre-commit` in my $PATH.  Is this related to the ups table or separate?

Comment (leeskelvin) [reply .pre-commit-config.yaml]: Thanks for catching this, autoupdate seemed to work well:
```
$ pre-commit autoupdate
Updating https://github.com/pre-commit/pre-commit-hooks ... [INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
updating v4.3.0 -> v4.4.0.
Updating https://github.com/psf/black ... [INFO] Initializing environment for https://github.com/psf/black.
updating 22.3.0 -> 23.1.0.
Updating https://github.com/PyCQA/isort ... [INFO] Initializing environment for https://github.com/PyCQA/isort.
updating 5.10.1 -> 5.12.0.
Updating https://github.com/PyCQA/flake8 ... [INFO] Initializing environment for https://github.com/PyCQA/flake8.
updating 4.0.1 -> 6.0.0.
```

**Comment by leeskelvin** [comment]
Updated `ups/source_injection.table` to include `setupRequired` entries for `daf_butler` and `pipe_base`.

File: .pre-commit-config.yaml
Comment (leeskelvin) [reply .pre-commit-config.yaml]: PS - `pre-commit` for me appears in my $PATH after setting up the stack in my working shell.

Comment (timj) [reply .pre-commit-config.yaml]: (you have to install `rubin-env-developer` to get it in your conda env if you aren't using a shared stack)

File: python/lsst/source/injection/inject_base.py
Comment (jmeyers314) [inline python/lsst/source/injection/inject_base.py:577]: Do we want to continue to support all of these in this refactor?  The only reasonable one (in my opinion) is star -> DeltaFunction.  I'd get rid of the rest.

File: python/lsst/source/injection/inject_base.py
Comment (jmeyers314) [inline python/lsst/source/injection/inject_base.py:596]: There's a lot more in the `galsim` namespace than just names of surface brightness profiles.  And even some things that are sbprofiles in an abstract sense won't work here because they're compound/derived (e.g., galsim.Sum, galsim.Deconvolution).  I think I'd just list the allowed profiles here for now: `['Gaussian', 'Moffat', 'Airy', 'Kolmogorov', 'Box', 'TopHat', 'Exponential', 'Sersic', 'Spergel', 'DeltaFunction', 'InclinedExponential', 'InclinedSersic', 'VonKarman']`

File: python/lsst/source/injection/inject_base.py
Comment (jmeyers314) [inline python/lsst/source/injection/inject_base.py:602]: Maybe unknown -> unsupported

File: python/lsst/source/injection/inject_engine.py
Comment (jmeyers314) [inline python/lsst/source/injection/inject_engine.py:138]: I think `Transformation` here is an implementation detail.  I'd just say this is a `galsim.gsobject.GSObject`.

File: python/lsst/source/injection/inject_engine.py
Comment (jmeyers314) [inline python/lsst/source/injection/inject_engine.py:147]: Fine for now, but I think this really depends on internal implementation details of GalSim to work (and I'm not sure if it always works or just usually works for all GSObjects in fact).  I could easily imagine a GSObject taking a kwargs with name `name` and using that to set an attr `.name2`.  I think the right way to approach this is actually to use something like https://galsim-developers.github.io/GalSim/_build/html/config_process.html#galsim.config.BuildGSObject.  Please make a issue to investigate.

File: python/lsst/source/injection/inject_engine.py
Comment (jmeyers314) [inline python/lsst/source/injection/inject_engine.py:240]: `100` feels really arbitrary to me.  What's the use case here?

**Review by jmeyers314 — CHANGES_REQUESTED** [review]
Hey Lee,
  Thanks for getting this all put together.  I have some concerns about the interface to GalSim here, but don't think they should hold up the current ticket as long as we're not locked in quite yet.  I left some suggestions.

File: doc/lsst.source.injection/reference/1_make_injection_pipeline.rst
Comment (leeskelvin) [reply doc/lsst.source.injection/reference/1_make_injection_pipeline.rst:67]: Yes, *but*, you'd need to also provide a butler repo for it to query against. This task at present requires no actual butler repo to exist, which makes it rather simple. We could remove the `-i` argument, but we'd need to replace it with a `-b` argument. I'm not sure which is more obvious?

Side note: you don't _have_ to specify `-i` in this command for it to succeed. If `-i` is not passed as an argument, this script will simply amend all the connection type names after `postISRCCD` to be prefixed with `injected_`. This may or may not be a desirable feature, but it is an option available at present.

File: doc/lsst.source.injection/reference/1_make_injection_pipeline.rst
Comment (leeskelvin) [reply doc/lsst.source.injection/reference/1_make_injection_pipeline.rst:109]: Similar comment to above - no butler is currently required here. Adding the automatic inference of the `injection_pipeline` would require that to exist on-disk already.

Another side note: the user may be querying against a completely new repo where `postISRCCD` has not yet been registered. In such a scenario, automatic inference of the injection pipeline would fail. This may be an edge case, but also something to consider.

File: doc/lsst.source.injection/reference/2_generate_injection_catalog.rst
Comment (leeskelvin) [reply doc/lsst.source.injection/reference/2_generate_injection_catalog.rst:213]: I've added a note on the user being able to provide their own catalog into this file, and fleshed this file out significantly. Changes have just been pushed and can be seen at [this link](https://s3df.slac.stanford.edu/people/lskelvin/inject/lsst.source.injection/reference/2_generate_injection_catalog.html).

Comment (leeskelvin) [reply doc/lsst.source.injection/reference/2_generate_injection_catalog.rst:213]: I'll attempt to finish up the docs to the best of my ability on this ticket, but I'm considering pulling the documentation out of this ticket and putting it onto another ticket for further review. Keen to hear your thoughts on that.

File: pipelines/inject_exposure.yaml
Comment (leeskelvin) [reply pipelines/inject_exposure.yaml]: YAML files updated to lowercase booleans.

File: python/lsst/source/injection/bin/generate_injection_catalog.py
Comment (leeskelvin) [reply python/lsst/source/injection/bin/generate_injection_catalog.py:139]: Modified the help string to read:
```
A string expression similar to an SQL WHERE clause to query for datasets with a WCS component.
```

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:148]: Two points come to mind:

1. During testing, whilst collaborating with another developer, I found I was not able to user their injection catalog that they sent to me as-is without needing to modify the `stamp` column. They had hard-coded the full path to these FITS files on their local disk, and I was not able to run with that catalog out of the box.
2. For large catalogs, an absolute file path encodes a lot of duplicated information. It becomes easier to parse the catalog if the `stamp` column only encodes the local stub, leaving the local root to be specified elsewhere.

With that said, there's also currently nothing stopping an end user from continuing to put the full path into the `stamp` column right now and leaving `stamp_prefix` as its default value (`""`). This will append nothing to the front of each filename, preserving the behavior you suggest.

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:160]: -> [DM-39721](https://jira.lsstcorp.org/browse/DM-39721)

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:576]: I'm split on this. On the one hand, keeping it in costs us little and may help legacy users in the conversion phase. However, as so much else has changed in catalog format/etc, then perhaps we don't need to keep this in. 

If you think it best, I'm happy to remove all other than:
```
"Star": "DeltaFunction"
```
(i.e., also throwing away the lowercase fixes and enforcing Pascal case formatting).

File: python/lsst/source/injection/inject_engine.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_engine.py:138]: Modified here and elsewhere in this file.

File: python/lsst/source/injection/inject_engine.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_engine.py:147]: :point_right: [DM-39725](https://jira.lsstcorp.org/browse/DM-39725)

File: python/lsst/source/injection/inject_engine.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_engine.py:240]: It is somewhat arbitrary. It's not beyond the scope of imagination that the first 10 HDUs may be filled with tabular data, but 1000 seemed too large. An order of magnitude argument led me to 100. Happy to modify as you see fit.

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:143]: Having spoken with member of AP who will be using this code base, I don't think it's trivial for them to define boolean selections based on catalog logic alone. They have template types that are defined elsewhere, and then that logic gets encoded into this boolean column.

I suppose we could replace it with your evaluation logic, and then AP could still include a column named `select` which could be selected against?

Let me look at implementing this, and I'll report back on this thread later today.

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:154]: This is a fairly important configuration option that was added in at request from several end-users. I worry that removing it will hamper our ability to pursue some of the things we want to do out of the gate until a more suitable approach comes up.

However, if we're happy to keep it in the short term, I'm happy to set up a ticket to look at replacing it?

Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:154]: PS - I'm also happy to remove the (perhaps redundant) `draw_size_scale` config option as we discussed last week, if that would be appropriate here?

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:576]: Replaced:
```
replace_dict = {
    "0": "DeltaFunction",
    "star": "DeltaFunction",
    "Star": "DeltaFunction",
    "1": "Sersic",
    "galaxy": "Sersic",
    "Galaxy": "Sersic",
    "2": "Trail",
    "trail": "Trail",
    "stamp": "Stamp",
}
```

with

```
replace_dict = {"Star": "DeltaFunction"}
```

File: doc/lsst.source.injection/reference/2_generate_injection_catalog.rst
Comment (leeskelvin) [reply doc/lsst.source.injection/reference/2_generate_injection_catalog.rst:213]: :point_right: [DM-39728](https://jira.lsstcorp.org/browse/DM-39728)

I'll remove documentation files from this PR prior to submission, and move them to DM-39728 instead for subsequent review.

File: python/lsst/source/injection/inject_engine.py
Comment (jmeyers314) [reply python/lsst/source/injection/inject_engine.py:147]: I think this should be close to what we want for determining arguments from a catalog row:

```python
def GetKwargs(row, cls):
    """Assemble dictionary of keyword arguments to use when constructing a 
    GSObject.
    """
    req = getattr(cls, '_req_params', {})
    opt = getattr(cls, '_opt_params', {})
    single = getattr(cls, '_single_params', {})

    params = {}

    # Check required args
    for key in req:
        if key not in row:
            raise ValueError(
                f"Required parameter {key} not found in input catalog"
            )
        params[key] = row[key]
    
    # Optional args
    for key in opt:
        if key in row:
            params[key] = row[key]
    
    # Single args
    for s in single:
        count = 0
        for key in s:
            if key in row:
                count += 1
                if count > 1:
                    raise ValueError(
                        f"Only one of {s.keys()} allowed for type {cls}"
                    )
                params[key] = row[key]
    if count == 0:
        raise ValueError(
            f"One of the args {s.keys()} is required for type {cls}"
        )
    return params
```

Comment (jmeyers314) [reply python/lsst/source/injection/inject_engine.py:147]: Note that most `GSObject`s will accept `flux` as an argument, so we might want to explicitly strip that out of the output of `GetKwargs` above and just apply `.withFlux()` after constructing the object.  Fortunately, I don't think it's wrong to set the flux in both places, just redundant.

File: doc/lsst.source.injection/reference/1_make_injection_pipeline.rst
Comment (leeskelvin) [reply doc/lsst.source.injection/reference/1_make_injection_pipeline.rst:74]: Following a discussion on Zoom about this today, I've opened this ticket: [DM-39748](https://jira.lsstcorp.org/browse/DM-39748). I think we can implement this change seamlessly, whilst still maintaining the current behavior. The ticket above will explore this in more detail.

File: doc/lsst.source.injection/reference/1_make_injection_pipeline.rst
Comment (leeskelvin) [reply doc/lsst.source.injection/reference/1_make_injection_pipeline.rst:116]: See above, :point_right: [DM-39748](https://jira.lsstcorp.org/browse/DM-39748)

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:143]: I think we're in agreement now that this can be implemented here and should maintain current behavior for legacy fakes users. The only issue is that end-users may need to know some of our internal names for certain objects. The flexibility granted via this change however is likely worth it, as long as this feature is well documented. 

The changes suggested above will be implemented on this ticket, and documentation ticket [DM-39728](https://jira.lsstcorp.org/browse/DM-39728) will document them.

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:154]: On discussion, we've agreed to leave `col_draw_size` in place for now (the config option which sets the name of the draw size column, default "draw_size"). This functionality is likely something that we'll want out of the box. We will however remove `draw_size_scale` and `draw_size_max` from this PR, as those are arguably unnecessary extras.

File: python/lsst/source/injection/inject_engine.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_engine.py:240]: On discussion, we'll remove this `while` loop and replace it with a `for` loop with an upper limit determined by astropy's FITS reader.

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py]: A new `selection` config option has now been added. This replaces the old `subset_by_select` on/off config option, and the `col_select` column name config option.

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py:142]: Both `draw_size_scale` and `draw_size_max` have been removed. The `col_draw_size` config remains, allowing the user to manually specify a draw size override.

File: python/lsst/source/injection/inject_engine.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_engine.py:234]: Two new free functions have now been implemented in `inject_engine`: `get_object_data` and `get_shear_data`. The former (`get_object_data`) is essentially what you have above, bar a few variable name changes and small implementation details/complete doc-string/etc.

I've also added `get_shear_data` which does the same thing but for the Shear kwargs. To achieve this, as we discussed, I've just hard-coded the possible list of allowed shear attributes into the function.

File: python/lsst/source/injection/inject_engine.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_engine.py:321]: Using `astropy.io.fits`, no `for` loop is actually required at all here. I've replaced this whole block with this:

```
    if os.path.exists(stamp_file):
        with fits.open(stamp_file) as hdul:
            hdu_images = [hdu.is_image and hdu.size > 0 for hdu in hdul]  # type: ignore
        if any(hdu_images):
            stamp_data = galsim.fits.read(stamp_file, read_header=True, hdu=np.where(hdu_images)[0][0])
        else:
            raise RuntimeError(f"Cannot find image in input FITS file {stamp_file}.")
```

That seems to nicely cover all bases; thanks for catching this!

**Comment by leeskelvin** [comment]
Thanks again for your comments above @jmeyers314. I think all are addressed, but please let me know if there's anything further of concern. Otherwise, will merge once you hit the approve button here.

**Comment by jmeyers314** [comment]
Hi Lee,
  I'm happy with all of this.  I'll leave it to you to decide what to do about the selection as a string config parameter.  Marking as reviewed now though.

**Comment by leeskelvin** [comment]
Thanks @jmeyers314. String selection has now been fully incorporated, including for per-visit selection. Will re-run Jenkins after stripping out the documentation scripts from above :+1:

File: python/lsst/source/injection/inject_base.py
Comment (leeskelvin) [reply python/lsst/source/injection/inject_base.py]: PS - this `selection` logic now also covers the prior visit selection logic as well.

**Review by jmeyers314 — APPROVED** [review]

**Pull request merged**
